File: 프로그래머스/3/43162. 네트워크/네트워크.java
import java.util.*;

class Solution {
    public void bfs(int x, int n, Map<Integer,List<Integer>> map, boolean[] v) {
        Queue<Integer> q = new ArrayDeque<>();
        q.add(x);
        v[x] = true;
        
        while (!q.isEmpty()) {
            int cur = q.poll();
            
            for(int l: map.getOrDefault(cur, new ArrayList<>())) {
                if (!v[l]) {
                    q.add(l);
                    v[l] = true;
                }
            }
        }
    }
    public int solution(int n, int[][] computers) {
        int answer = 0;
        
        Map<Integer,List<Integer>> map = new HashMap<>();
        for(int i=0;i<n;i++) {
            for(int j=0;j<n;j++) {
                if (i!=j && computers[i][j] == 1) {
                    map.computeIfAbsent(i, k->new ArrayList<>()).add(j);
                }
            }
        }
        
        boolean[] visited = new boolean[n];
        for(int i=0;i<n;i++) {
            if(!visited[i]) {
                bfs(i,n,map,visited);
                answer++;
            }
        }
        
        return answer;
    }
}

File: 프로그래머스/3/43162. 네트워크/네트워크.swift
import Foundation

func solution(_ n:Int, _ computers:[[Int]]) -> Int {
    var grid:[Int: [Int]] = [:] 
    for i in 0..<n {
        for j in i+1..<n {
            if computers[i][j] == 1 {
                grid[i+1, default:[]].append(j+1)
                grid[j+1, default:[]].append(i+1)
            }
        }
    }
    
    var visited = Array(repeating: false, count: n+1)
    
    func bfs(_ st: Int) {
        var q = [st]
        var head = 0
        visited[st] = true
        
        while head < q.count {
            let cur = q[head]
            head += 1
            
            if let nexts = grid[cur]{
                for nxt in nexts {
                    if !visited[nxt] {
                        q.append(nxt)
                        visited[nxt] = true
                    }
                }
            }
        }
    }
    
    var answer = 0
    
    for i in 1...n {
        if !visited[i] {
            bfs(i)
            answer += 1
        }
    }
    
    
    return answer
}